ruby_event_store/spec/in_memory_repository_spec.rb

```diff
@@ -12,6 +12,28 @@ module RubyEventStore
 
     it_behaves_like :event_repository, InMemoryRepository
 
+    it 'does not allow same event twice in a stream - checks stream events before checking all events' do
+      repository = InMemoryRepository.new
+      repository.append_to_stream(
+        TestDomainEvent.new(event_id: eid = "fbce0b3d-40e3-4d1d-90a1-901f1ded5a4a"),
+        'other',
+        -1
+      )
+      repository.append_to_stream(
+        TestDomainEvent.new(event_id: "a1b49edb-7636-416f-874a-88f94b859bef"),
+        'stream',
+        -1
+      )
+      expect(eid).not_to receive(:eql?)
```

Comment:
Is this expectation necessary? Looks like a detail of the implementation. What was the intent here?

Reply:
Let's look at the bigger picture here:

```
    def link_to_stream(event_ids, stream_name, expected_version)
      events = normalize_to_array(event_ids).map{|eid| read_event(eid) }
      add_to_stream(events, expected_version, stream_name, nil)
    end

    def append(events, expected_version, stream, stream_name, include_global)
      raise WrongExpectedEventVersion unless (stream.size - 1).equal?(expected_version)
      events.each do |event|
        raise EventDuplicatedInStream if stream.any?{|ev| ev.event_id.eql?(event.event_id) }
        if include_global
          raise EventDuplicatedInStream if all.any?{|ev| ev.event_id.eql?(event.event_id) }
          all.push(event)
        end
        stream.push(event)
      end
      streams[stream_name] = stream
      self
    end
```

* mutant told me that I could change the `raise EventDuplicatedInStream if stream.any?{|ev| ev.event_id.eql?(event.event_id) }` into `equal?`.
* I was surprised by that :-) and I wondered why. I turned out that the reason is that because when we are linking events we first read them from in memory so `equal?` is good enough (same in memory string is compared). When we append it's different string so `equal?` is not good enough but then  `include_global` is `true` and the next line checks it so everything is ok.
* I had 2 choices.
    * change `eql` to `equal` which I did not like
    * write a test showing why `eql` a not `equal`. I chose this option and this is the test.

It is indeed an implementation detail and that's why it is in a repository test and not in a linter.

The intention of this test is to prove that we check for duplicates in events of a single stream before checking all events. One might call it a performance test (shorter collection is checked before longer collection is checked). The way it is done is by ensuring that we don't compare equality with the other event from the other stream.

Alternative solution:

* disable mutation testing in a single method
* use `equal` instead of `eql` - even though I would prefer not to.
* moving checks into higher methods maybe...
* something else I haven't thought about

That's the story more or less. What's your take on it?